### ISSUE TITLE
[SwiftPM][Diagnostics] Missing or misplaced `\u{1b}[2K` sequence in build output

### ISSUE THREAD

**Issue by philipturner** [apple/swift#58916]
**Describe the bug**
I am changing [Swift-Colab](https://github.com/philipturner/swift-colab) to use `pexpect` instead of `subprocess.Popen` to launch `swift build` in Swift-Colab. It now handles output lines differently, paying attention to ANSI escape sequences like `\u{1b}[2K` that clear a line and replace it with something new. This decreases the size of build logs from SwiftPM, but there may be a glitch with how the compiler places these escape sequences. This causes stray colons and periods to appear in output. I replicated this bug using the inline Bash functionality of Colab in Python mode, so it's not a fault of Swift-Colab itself. I have not been able to reproduce this on macOS.

Before (Swift-Colab 1.0 and 2.0)
```
Installing packages:
	.package(url: "https://github.com/KarthikRIyer/swiftplot", .branch("master"))
		SwiftPlot
		AGGRenderer
With SwiftPM flags: [-Xcc, -isystem/usr/include/freetype2, -Xswiftc, -lfreetype]
Working in: /tmp/tmpboleba0t/swift-install
Fetching https://github.com/KarthikRIyer/swiftplot
Fetched https://github.com/KarthikRIyer/swiftplot (1.38s)
Creating working copy for https://github.com/KarthikRIyer/swiftplot
Working copy of https://github.com/KarthikRIyer/swiftplot resolved at master
[1/77] Compiling lodepng lodepng.cpp
[2/77] Compiling CPPAGGRenderer CPPAGGRenderer.cpp
While building module 'CPPAGGRenderer' imported from /tmp/tmpboleba0t/swift-install/package/.build/checkouts/swiftplot/Sources/AGGRenderer/CAGGRenderer/CAGGRenderer.cpp:2:
In file included from <module-includes>:1:
/tmp/tmpboleba0t/swift-install/package/.build/checkouts/swiftplot/Sources/AGGRenderer/CPPAGGRenderer/include/CPPAGGRenderer.h:38:1: warning: umbrella header for module 'CPPAGGRenderer' does not include header 'savebmp.h' [-Wincomplete-umbrella]
/tmp/tmpboleba0t/swift-install/package/.build/checkouts/swiftplot/Sources/AGGRenderer/CPPAGGRenderer/include/CPPAGGRenderer.h:38:1: warning: umbrella header for module 'CPPAGGRenderer' does not include header 'pixel_formats.h' [-Wincomplete-umbrella]
2 warnings generated.
[3/77] Compiling CAGGRenderer CAGGRenderer.cpp
[4/77] Compiling AGG gpc.c
[5/77] Compiling agg_spline_ctrl.cpp
[6/77] Compiling agg_slider_ctrl.cpp
[7/77] Compiling agg_scale_ctrl.cpp
[8/77] Compiling agg_rbox_ctrl.cpp
...
```

After (Swift-Colab 2.1):
```
Installing package:
    .package(url: "https://github.com/KarthikRIyer/swiftplot", .branch("master"))
        SwiftPlot
        AGGRenderer
With SwiftPM flags: ["-Xcc", "-isystem/usr/include/freetype2", "-Xswiftc", "-lfreetype"]
Working in: /opt/swift/packages
Fetching https://github.com/KarthikRIyer/swiftplot
[6688/6688] Fetching objects
Fetched https://github.com/KarthikRIyer/swiftplot (2.08s)
Creating working copy for https://github.com/KarthikRIyer/swiftplot
Working copy of https://github.com/KarthikRIyer/swiftplot resolved at master
Building for debugging...
:
:
umbrella header for module 'CPPAGGRenderer' does not include header 'savebmp.h' [-Wincomplete-umbrella]
umbrella header for module 'CPPAGGRenderer' does not include header 'pixel_formats.h' [-Wincomplete-umbrella]
.
/opt/swift/packages/2/.build/checkouts/swiftplot/Sources/SwiftPlot/Histogram.swift:283:21: warning: initialization of immutable value 'series' was never used; consider replacing with assignment to '_' or removing it
                let series = allSeries[seriesIdx]
                ~~~~^~~~~~
                _
```

**How to reproduce**

Here are instructions for how to reproduce the bug via Python mode. For Swift mode (the easier way), just run the [SwiftPlot](https://github.com/philipturner/swift-colab#testing) test notebook in Swift-Colab's README.

1. Make a Colab notebook and only install SwiftPlot as a dependency. To do this, make the following cell and run it once. SwiftPlot will build without errors.
```swift
%install-swiftpm-flags $clear 
// %install-swiftpm-flags -Xcc -isystem/usr/include/freetype2 -Xswiftc -lfreetype
%install '.package(url: "https://github.com/KarthikRIyer/swiftplot", .branch("master"))' SwiftPlot AGGRenderer
```
2. Then make another cell, which will allow you to switch from Swift to Python mode. Run it.
```swift
%system echo "python" > "/opt/swift/runtime"
```
3. In the GUI, go to the "Runtime" tab at the top. Click "Restart runtime".
4. Make another cell. You're now in Python mode. This uses Bash commands to repeat the build process, but modifies it by adding the flags `-c release -Xswiftc -Onone`.
```
!cd /opt/swift/packages/1 && /opt/swift/toolchain/usr/bin/swift-build -c release -Xswiftc -Onone -Xcc -isystem/usr/include/freetype2 -Xswiftc -lfreetype
```

**Explanation**

My guess is it's misplacing or missing a clear escape sequence at the end of the lines. Regarding the 5 lines that look off (`:`, `:`, `umbrella...`, `umbrella...`, `.`), they are mirrored by 5 lines in the output in Swift-Colab 2.0 (`While...`, `In...`, `/tmp...`, `/tmp...`, `2...`). They usually constitute the last component of a line after a number or colon. I got the raw binary output of the Stdout, and turned it into an array of `UInt8`. I ran it through the Swift REPL to convert it back into characters, and found something interesting.

Output for four lines: `:`, `:`, `umbrella...`, `umbrella...`

<details>
<summary>Here</summary>

```
["\u{1B}", "[", "2", "K", "\r", "W", "h", "i", "l", "e", " ", "b", "u", "i", "l", "d", "i", "n", "g", " ", "m", "o", "d", "u", "l", "e", " ", "\'", "\u{1B}", "[", "2", "K", "\r", "C", "P", "P", "A", "G", "G", "R", "e", "n", "d", "e", "r", "e", "r", "\u{1B}", "[", "2", "K", "\r", "\'", " ", "i", "m", "p", "o", "r", "t", "e", "d", " ", "f", "r", "o", "m", " ", "\u{1B}", "[", "2", "K", "\r", "/", "o", "p", "t", "/", "s", "w", "i", "f", "t", "/", "p", "a", "c", "k", "a", "g", "e", "s", "/", "2", "/", ".", "b", "u", "i", "l", "d", "/", "c", "h", "e", "c", "k", "o", "u", "t", "s", "/", "s", "w", "i", "f", "t", "p", "l", "o", "t", "/", "S", "o", "u", "r", "c", "e", "s", "/", "A", "G", "G", "R", "e", "n", "d", "e", "r", "e", "r", "/", "C", "A", "G", "G", "R", "e", "n", "d", "e", "r", "e", "r", "/", "C", "A", "G", "G", "R", "e", "n", "d", "e", "r", "e", "r", ".", "c", "p", "p", "\u{1B}", "[", "2", "K", "\r", ":", "\u{1B}", "[", "2", "K", "\r", "2", "\u{1B}", "[", "2", "K", "\r", ":", "\r", "\n", "\u{1B}", "[", "2", "K", "\r", "I", "n", " ", "f", "i", "l", "e", " ", "i", "n", "c", "l", "u", "d", "e", "d", " ", "f", "r", "o", "m", " ", "\u{1B}", "[", "2", "K", "\r", "<", "m", "o", "d", "u", "l", "e", "-", "i", "n", "c", "l", "u", "d", "e", "s", ">", "\u{1B}", "[", "2", "K", "\r", ":", "\u{1B}", "[", "2", "K", "\r", "1", "\u{1B}", "[", "2", "K", "\r", ":", "\r", "\n", "\u{1B}", "[", "2", "K", "\r", "/", "o", "p", "t", "/", "s", "w", "i", "f", "t", "/", "p", "a", "c", "k", "a", "g", "e", "s", "/", "2", "/", ".", "b", "u", "i", "l", "d", "/", "c", "h", "e", "c", "k", "o", "u", "t", "s", "/", "s", "w", "i", "f", "t", "p", "l", "o", "t", "/", "S", "o", "u", "r", "c", "e", "s", "/", "A", "G", "G", "R", "e", "n", "d", "e", "r", "e", "r", "/", "C", "P", "P", "A", "G", "G", "R", "e", "n", "d", "e", "r", "e", "r", "/", "i", "n", "c", "l", "u", "d", "e", "/", "C", "P", "P", "A", "G", "G", "R", "e", "n", "d", "e", "r", "e", "r", ".", "h", "\u{1B}", "[", "2", "K", "\r", ":", "\u{1B}", "[", "2", "K", "\r", "3", "8", "\u{1B}", "[", "2", "K", "\r", ":", "\u{1B}", "[", "2", "K", "\r", "1", "\u{1B}", "[", "2", "K", "\r", ":", "\u{1B}", "[", "2", "K", "\r", " ", "\u{1B}", "[", "2", "K", "\r", "w", "a", "r", "n", "i", "n", "g", ":", " ", "\u{1B}", "[", "2", "K", "\r", "u", "m", "b", "r", "e", "l", "l", "a", " ", "h", "e", "a", "d", "e", "r", " ", "f", "o", "r", " ", "m", "o", "d", "u", "l", "e", " ", "\'", "C", "P", "P", "A", "G", "G", "R", "e", "n", "d", "e", "r", "e", "r", "\'", " ", "d", "o", "e", "s", " ", "n", "o", "t", " ", "i", "n", "c", "l", "u", "d", "e", " ", "h", "e", "a", "d", "e", "r", " ", "\'", "s", "a", "v", "e", "b", "m", "p", ".", "h", "\'", " ", "[", "-", "W", "i", "n", "c", "o", "m", "p", "l", "e", "t", "e", "-", "u", "m", "b", "r", "e", "l", "l", "a", "]", "\u{1B}", "[", "2", "K", "\r", "\r", "\n", "\u{1B}", "[", "2", "K", "\r", "/", "o", "p", "t", "/", "s", "w", "i", "f", "t", "/", "p", "a", "c", "k", "a", "g", "e", "s", "/", "2", "/", ".", "b", "u", "i", "l", "d", "/", "c", "h", "e", "c", "k", "o", "u", "t", "s", "/", "s", "w", "i", "f", "t", "p", "l", "o", "t", "/", "S", "o", "u", "r", "c", "e", "s", "/", "A", "G", "G", "R", "e", "n", "d", "e", "r", "e", "r", "/", "C", "P", "P", "A", "G", "G", "R", "e", "n", "d", "e", "r", "e", "r", "/", "i", "n", "c", "l", "u", "d", "e", "/", "C", "P", "P", "A", "G", "G", "R", "e", "n", "d", "e", "r", "e", "r", ".", "h", "\u{1B}", "[", "2", "K", "\r", ":", "\u{1B}", "[", "2", "K", "\r", "3", "8", "\u{1B}", "[", "2", "K", "\r", ":", "\u{1B}", "[", "2", "K", "\r", "1", "\u{1B}", "[", "2", "K", "\r", ":", "\u{1B}", "[", "2", "K", "\r", " ", "\u{1B}", "[", "2", "K", "\r", "w", "a", "r", "n", "i", "n", "g", ":", " ", "\u{1B}", "[", "2", "K", "\r", "u", "m", "b", "r", "e", "l", "l", "a", " ", "h", "e", "a", "d", "e", "r", " ", "f", "o", "r", " ", "m", "o", "d", "u", "l", "e", " ", "\'", "C", "P", "P", "A", "G", "G", "R", "e", "n", "d", "e", "r", "e", "r", "\'", " ", "d", "o", "e", "s", " ", "n", "o", "t", " ", "i", "n", "c", "l", "u", "d", "e", " ", "h", "e", "a", "d", "e", "r", " ", "\'", "p", "i", "x", "e", "l", "_", "f", "o", "r", "m", "a", "t", "s", ".", "h", "\'", " ", "[", "-", "W", "i", "n", "c", "o", "m", "p", "l", "e", "t", "e", "-", "u", "m", "b", "r", "e", "l", "l", "a", "]", "\u{1B}", "[", "2", "K", "\r", "\r", "\n"]
```

</details>

Pay attention to the very end of the first line (a `:` line) and the very beginning of the next line (also a line that becomes just `:` in Swift-Colab 2.1). I extracted it from the raw output. As a string, it is this:
```
:2:
In
```
As bytes, it is this:
```
":", "\u{1B}", "[", "2", "K", "\r", "2", "\u{1B}", "[", "2", "K", "\r", ":", "\r", "\n", "\u{1B}", "[", "2", "K", "\r", "I", "n"
```
Notice that there is no `\u{1B}[2K` between the last colon and the `\r\n` after it. I hypothesize that there should be, and that's what's causing the bug.

**Other remarks**

I found this in Swift-Colab, but it should be possible to reproduce anywhere. Perhaps it's specific to the Linux that runs on Google Colaboratory (Ubuntu 18.04), or it happens on Linux somewhere else. The Python Bash functionality is not something I made; it's part of the IPython repo and like a terminal app. Thus, if someone else compiled SwiftPlot on Linux via command-line, they would get the same problem. This should happen when compiling any Clang module with that umbrella header error. Thus, it is not specific to SwiftPlot either.

**Comment by philipturner** [comment]
CC: @KarthikRIyer

**Comment by philipturner** [comment]
This also creates some nice text colorization. How does this problem not occur in the macOS Terminal app?

<img width="787" alt="Screen Shot 2022-06-05 at 6 57 46 PM" src="https://user-images.githubusercontent.com/71743241/172074099-4a0a7ffa-dc06-4c58-96ff-04af96c7e1ca.png">

**Comment by philipturner** [comment]
This seems strangely fixed when I built SwiftPlot on the May 11, 2022 toolchain. I need to investigate further.

**Comment by philipturner** [comment]
Problem (5) in https://github.com/philipturner/swift-colab/issues/14 was the source of this bug report. Problem (1) originated from the same bug and disappeared in recent toolchains.

**Comment by philipturner** [comment]
The fix also went into the Swift 5.6.2 release toolchain. It seems like a bug fix was cherry-picked from `apple:main` between 5.6.1 and 5.6.2.

**Comment by philipturner** [comment]
It looks like the bug is not fixed on the Swift 5.6.2 release toolchain:

```
Installing package:
    .package(url: "https://github.com/KarthikRIyer/swiftplot", .branch("master"))
        SwiftPlot
        AGGRenderer
With SwiftPM flags: []
Working in: /opt/swift/packages
Fetching https://github.com/KarthikRIyer/swiftplot
[6703/6703] Fetching objects
Fetched https://github.com/KarthikRIyer/swiftplot (2.40s)
Creating working copy for https://github.com/KarthikRIyer/swiftplot
Working copy of https://github.com/KarthikRIyer/swiftplot resolved at master
Building for debugging...
:
:
umbrella header for module 'CPPAGGRenderer' does not include header 'savebmp.h' [-Wincomplete-umbrella]
umbrella header for module 'CPPAGGRenderer' does not include header 'pixel_formats.h' [-Wincomplete-umbrella]
2 warnings generated.
```

But it is fixed on the 2022-08-09 snapshot:

```
Installing package:
    .package(url: "https://github.com/KarthikRIyer/swiftplot", .branch("master"))
        SwiftPlot
        AGGRenderer
With SwiftPM flags: []
Working in: /opt/swift/packages
Fetching https://github.com/KarthikRIyer/swiftplot
Fetched https://github.com/KarthikRIyer/swiftplot (2.12s)
Creating working copy for https://github.com/KarthikRIyer/swiftplot
Working copy of https://github.com/KarthikRIyer/swiftplot resolved at master
Building for debugging...
While building module 'CPPAGGRenderer' imported from /opt/swift/packages/1/.build/checkouts/swiftplot/Sources/AGGRenderer/CAGGRenderer/CAGGRenderer.cpp:2:
In file included from <module-includes>:1:
/opt/swift/packages/1/.build/checkouts/swiftplot/Sources/AGGRenderer/CPPAGGRenderer/include/CPPAGGRenderer.h:38:1: warning: umbrella header for module 'CPPAGGRenderer' does not include header 'savebmp.h' [-Wincomplete-umbrella]
/opt/swift/packages/1/.build/checkouts/swiftplot/Sources/AGGRenderer/CPPAGGRenderer/include/CPPAGGRenderer.h:38:1: warning: umbrella header for module 'CPPAGGRenderer' does not include header 'pixel_formats.h' [-Wincomplete-umbrella]
2 warnings generated.
[88/88] Linking libjupyterInstalledPackages1.so
Build complete! (32.61s)
=== ------------------------------------------------------------------------ ===
=== The following Clang modules cannot be imported in your source code until ===
=== you restart the runtime. If you do not intend to explicitly import       ===
=== modules listed here, ignore this warning.                                ===
=== ["lodepng", "CFreeType", "CPPAGGRenderer", "AGG", "CAGGRenderer"]
=== ------------------------------------------------------------------------ ===
```

Both are using the latest release of Swift-Colab. Therefore, I will mark this issue as resolved.
